src/cdk/a11y/input-modality/input-modality-detector.ts

```diff
@@ -0,0 +1,187 @@
+  ignoreKeys: ['Alt', 'Control', 'Meta', 'Shift'],
+};
+
+/**
+ * The amount of time needed to pass after a touchstart event in order for a subsequent mousedown
+ * event to be attributed as mouse and not touch.
+ */
+export const TIME_AFTER_TOUCH_MS = 650;
+
+/**
+ * Event listener options that enable capturing and also mark the listener as passive if the browser
+ * supports it.
+ */
+const opts = normalizePassiveListenerOptions({
+  passive: true,
+  capture: true,
+});
+
+/**
+ * Service that detects the user's input modality.
+ *
+ * Note: This service will not update the input modality when a user is navigating with a screen
+ * reader (e.g. linear navigation with VoiceOver, object navigation / browse mode with NVDA, virtual
+ * PC cursor mode with JAWS). This is in part due to technical limitations (i.e. keyboard events
+ * do not fire as expected in these modes) but is also arguably the correct behavior. Navigating
+ * with a screen reader is akin to visually scanning a page, and should not be interpreted as actual
+ * user input interaction.
+ *
+ * When a user is not navigating but *interacting* with a screen reader, this service's behavior is
+ * largely undefined and depends on the events fired. For example, in VoiceOver, no keyboard events
+ * are fired when performing an element's default action via Caps Lock + Space, thus no input
+ * modality is detected.
+ */
+@Injectable({ providedIn: 'root' })
+export class InputModalityDetector implements OnDestroy {
+  /** Emits when the input modality changes. */
+  readonly inputModalityChange: Observable<InputModality>;
+
+  /** Returns the last detected / current input modality. */
+  get inputModality(): InputModality {
+    return this._inputModality;
+  }
+
+  /** The underlying ReplaySubject that emits input modality changes. */
+  private readonly _change = new ReplaySubject<InputModality>(1);
+
+  /** The current / last detected input modality by this service. */
+  private _inputModality: InputModality = null;
+
+  /** Options for this InputModalityDetector. */
+  private readonly _options: InputModalityDetectorOptions;
+
+  /**
+   * The timestamp of the last touch input modality. Used to determine whether mousedown events
+   * should be attributed to mouse or touch.
+   */
+  private _lastTouchMs = 0;
+
+  /**
+   * Handles keyboard events. Must be an arrow function in order to preserve the context when it
+   * gets bound.
+   */
+  private _onKeydown = (event: KeyboardEvent) => {
+    // If this is one of the keys we should ignore, then ignore it and don't update the input
+    // modality to keyboard.
+    if (this._options?.ignoreKeys?.some(key => key === event.key)) { return; }
+
+    this._updateInputModality('keyboard');
+  }
+
+  /**
+   * Handles mouse events. Must be an arrow function in order to preserve the context when it gets
+   * bound.
+   */
+  private _onMousedown = (event: MouseEvent) => {
+    if (isFakeMousedownFromScreenReader(event)) { return; }
```

Comment:
nit: IMO the early `return` here and in `_onTouchStart` makes things harder to follow, because there's so little logic after it. It might be clearer to just wrap the rest of the logic in an `if`.

Reply:
I almost always prefer returning early to reduce nesting. I'll leave this as-is for now, but if Jeremy/Miles/anyone has strong opinions either way, I'll switch to your recommendation.